### PR TITLE
Add manifests subscriptions

### PR DESCRIPTION
**Pull request by wilsonianb**
Add verify method to manifest script to check signature

### CODE REVIEW

**Comment by vinniefalco** [comment]
Nice

**Comment by vinniefalco** [comment]
Unfortunately this increases the number of lines which are not covered in tests, can we do something about it?

**Comment by wilsonianb** [comment]
I can at least add `Manifest::getSignature` to https://github.com/ripple/rippled/blob/develop/src/ripple/overlay/tests/manifest_test.cpp

For this and https://github.com/ripple/rippled/pull/1422, are there any existing websocket subscription tests I can reference?

**Comment by vinniefalco** [comment]
@wilsonianb Unfortunately we can't test websockets from C++ yet

File: src/ripple/overlay/tests/manifest_test.cpp [hunk 1]
```diff
@@ -179,6 +179,26 @@ class manifest_test : public ripple::TestSuite
...
+        STObject st(sfGeneric);
+        set(st, sfSequence, 0);
+        set(st, sfPublicKey, pk);
+        set(st, sfSigningPubKey, kp.first);
+        Serializer ss;
+        ss.add32(HashPrefix::manifest);
+        st.addWithoutSigningFields(ss);
+        auto const sig = sign(KeyType::ed25519, sk, ss.slice());
+
+        expect (strHex(sig) == strHex(m.getSignature()));
```
Comment (seelabs) [inline src/ripple/overlay/tests/manifest_test.cpp]: Missing `#include <ripple/basics/StringUtilities.h>` needed for `strHex` (or the nonunity build fails)

File: bin/python/ripple/util/Sign.py [hunk 2]
```diff
@@ -113,6 +116,16 @@ def get_signature(seq, validator_public_key_human, private_key_human):
...
+def verify_signature(seq, validator_public_key_human, public_key_human, signature):
+    v, validator_public_key = Base58.decode_version(validator_public_key_human)
+    check_validator_public(v, validator_public_key)
+
+    v, public_key = Base58.decode_version(public_key_human)
+
+    m = make_manifest(public_key, validator_public_key, seq)
+    public_key = public_key[1:]     # Remove ED25519_BYTE
+    sig = signature.decode('hex')
+    ed25519.checkvalid(sig, 'MAN\0' + m, public_key)
```
Comment (seelabs) [inline bin/python/ripple/util/Sign.py]: Add a `test_verify_signature` function to `test_Sign.py` unit test this function. FYI, to run the python unit tests, go to `bin/python` and run `python -m unittest discover`

File: src/ripple/app/misc/NetworkOPs.cpp [hunk 3]
```diff
@@ -1507,6 +1512,38 @@ void NetworkOPsImp::consensusViewChange ()
...
+        jvObj [jss::master_key]  = toBase58(TokenType::TOKEN_NODE_PUBLIC, mo.masterKey);
+        jvObj [jss::signing_key] = toBase58(TokenType::TOKEN_NODE_PUBLIC, mo.signingKey);
+        jvObj [jss::seq]         = Json::UInt (mo.sequence);
+        jvObj [jss::signature]   = strHex (mo.getSignature ());
+
+        for (auto i = mSubManifests.begin (); i != mSubManifests.end (); )
+        {
+            InfoSub::pointer p = i->second.lock ();
+
+            if (p)
```
Comment (seelabs) [inline src/ripple/app/misc/NetworkOPs.cpp]: Nit: I prefer `if (auto p = i->second.lock())` to reduce the scope of `p` to where it's needed. This is a nit, you can leave the code as is if you want.

File: src/ripple/overlay/impl/OverlayImpl.cpp [hunk 4]
```diff
@@ -695,6 +696,10 @@ OverlayImpl::onManifests (
                 app_.validators(),
                 journal);
 
+            if (result == ManifestDisposition::accepted ||
+                    result == ManifestDisposition::untrusted)
+                app_.getOPs().pubManifest (std::move(*make_Manifest(serialized)));
```
Comment (seelabs) [inline src/ripple/overlay/impl/OverlayImpl.cpp]: `std::move` is not needed here, `*make_Manifest` will already be an rvalue ref. Plus the parameter to `pubManifest` is a `Manifest const&`, so it's not going to take advantage of move semantics anyway.

File: src/ripple/overlay/tests/manifest_test.cpp [hunk 5]
```diff
@@ -179,6 +179,26 @@ class manifest_test : public ripple::TestSuite
...
+        testcase ("getSignature");
+        auto const sk = randomSecretKey();
+        auto const pk = derivePublicKey(KeyType::ed25519, sk);
+        auto const kp = randomKeyPair(KeyType::secp256k1);
+        auto const m = make_Manifest (KeyType::ed25519, sk, kp.first, 0);
+
+        STObject st(sfGeneric);
+        set(st, sfSequence, 0);
+        set(st, sfPublicKey, pk);
+        set(st, sfSigningPubKey, kp.first);
```
Comment (vinniefalco) [inline src/ripple/overlay/tests/manifest_test.cpp]: `st[sfPublicKey] = pk` should also work

Comment (wilsonianb) [inline src/ripple/overlay/tests/manifest_test.cpp]: Changed

File: bin/python/ripple/util/Sign.py (hunk 2 above)
Comment (wilsonianb) [inline bin/python/ripple/util/Sign.py]: Added python tests

File: src/ripple/app/misc/NetworkOPs.cpp (hunk 3 above)
Comment (wilsonianb) [inline src/ripple/app/misc/NetworkOPs.cpp]: Changed

File: src/ripple/overlay/impl/OverlayImpl.cpp (hunk 4 above)
Comment (wilsonianb) [inline src/ripple/overlay/impl/OverlayImpl.cpp]: Fixed

**Comment by seelabs** [comment]
The code is good, and I'm ready to sign off on that. However, I don't see any test that actually verifies that manifest are sent to the subscriber (I do see a test that a `subscribe` request returns success). How can I test this?

**Comment by wilsonianb** [comment]
@seelabs I was testing by peering with the altnet and then rotating one of the altnet validator's ephemeral keys.

My understanding is that I can't really test this in C++ at this time.

Do you know if it would be possible for me to submit a `mtMANIFESTS` message to the rippled in the `subscribe manifests` JavaScript test?
https://github.com/wilsonianb/rippled/blob/subscribe-manifests/test/jsonrpc-test.js#L173-L186

**Comment by vinniefalco** [comment]
@wilsonianb No, because `mtMANIFESTS` is a peer protocol message and not part of the client API. You could add it though. But it would be a facility purely for testing. Although in the future, it might prove useful to thin clients.

**Comment by seelabs** [comment]
:+1:

**Comment by ximinez** [comment]
Merged 749b4adc7c9a11d1d94295274b79fe80f1bf8055

**Pull request closed without merging**